lib/diff.ml

```diff
@@ -125,37 +145,153 @@ let rec type_item ~typing_env ~name ~reference ~current =
+              | _ ->
+                  Some
+                    (Type
+                       {
+                         tname = name;
+                         tdiff = Modified (Record_diff changed_lbls);
+                       }))
+          | ( Type_variant (ref_constructor_lst, _),
+              Type_variant (cur_constructor_lst, _) ) -> (
+              let changed_constrs =
+                modified_variant_type ~typing_env ~ref_constructor_lst
+                  ~cur_constructor_lst
+              in
+              match changed_constrs with
+              | [] -> None
+              | _ ->
+                  Some
+                    (Type
+                       {
+                         tname = name;
+                         tdiff = Modified (Variant_diff changed_constrs);
+                       }))
+          | _ ->
               Some
                 (Type
                    {
                      tname = name;
                      tdiff = Modified (Atomic { reference; current });
                    })))
 
-and modified_record_type ~typing_env ~ref_label_lst ~cur_label_lst =
+and modified_variant_type ~typing_env ~ref_constructor_lst ~cur_constructor_lst
+    =
+  let diff_cstrs name cstr1 cstr2 =
+    match (cstr1.cd_args, cstr2.cd_args) with
+    | Cstr_tuple type_lst1, Cstr_tuple type_lst2
+      when not
+             ((List.length type_lst1 = 1 && List.length type_lst2 = 1)
+             && not
+                  (Ctype.does_match typing_env (List.hd type_lst1)
+                     (List.hd type_lst2))) ->
+        let tuple_diff = modified_tuple_type ~typing_env type_lst1 type_lst2 in
+        if
+          List.for_all
+            (fun t ->
+              match t with Either.Left _ -> true | Either.Right _ -> false)
+            tuple_diff
+        then None
+        else Some { csname = name; csdiff = Modified (Tuple_c tuple_diff) }
+    | Cstr_record ref_label_lst, Cstr_record cur_label_lst ->
+        let record_diff =
+          modified_record_type ~typing_env ~ref_label_lst ~cur_label_lst
+        in
+        if List.length record_diff = 0 then None
+        else Some { csname = name; csdiff = Modified (Record_c record_diff) }
+    | _ ->
+        Some
+          {
+            csname = name;
+            csdiff = Modified (Atomic_c { reference = cstr1; current = cstr2 });
+          }
+  in
+  let ref_cstrs = extract_cstrs ref_constructor_lst in
+  let curr_cstrs = extract_cstrs cur_constructor_lst in
+  let modified_cstrs =
+    String_map.merge
+      (fun name ref cur ->
+        match (ref, cur) with
+        | None, None -> None
+        | Some ref, None -> Some { csname = name; csdiff = Removed ref }
+        | None, Some cur -> Some { csname = name; csdiff = Added cur }
+        | Some ref, Some cur -> diff_cstrs name ref cur)
+      ref_cstrs curr_cstrs
+    |> String_map.bindings |> List.map snd
+  in
+  modified_cstrs
+
+and modified_tuple_type ~typing_env (ref_tuple : type_expr list)
```

Comment:
This function can be simplified a lot. I think we'll run into similar cases in the future so I suggest extracting a generic:
```ocaml
val diff_list :
  dif_one: (ref: 'a option -> current: 'a option -> ('a, 'diff) Either.t) ->
  ref: 'a list ->
  current: 'a list ->
  ('a, 'diff) Either.t list
```
Where `diff_one` would receive `None` when there is no corresponding element in the relevant list because their length differ.

This can be implemented in a single traversal of both list. Let me know if you need help writing this.

Once you have that generic function, writing `modified_tuple_type` should be trivial.